### PR TITLE
Py-Libensemble: Add E4S testsuite stand alone test 

### PR DESCRIPTION
**Pull request by RikkiButler20**
This PR represents a preliminary smoke test for the `py-libensemble` package and is intended to serve the following purposes:

1. Leverage the current E4S test suite for the software
2. Demonstrate to package maintainers how to start a Spack stand alone test.

### CODE REVIEW

**Comment by shuds13** [comment]
I am happy to bring in as it does not break anything.

**Comment by tldahlgren** [comment]
@spackbot run pipeline

**Comment by shuds13** [comment]
This test would be run from the command line as:

`python tutorial_calling.py --comms local --nworkers 4`

I have tried this locally, modifying the code to:

```python
        run_cmd = [exe] + "--comms local --nworkers 4".split()

        self.run_test(exe='python',
                      options=run_cmd,
                      purpose='test: run {0} example'.format(exe),
                      work_dir=test_dir)
```

and running:

`spack install --test=root --overwrite -v -y py-libensemble`

but I get a message I don't understand

`Warning: RUN-TESTS: method not implemented [test]`

and I dont think the test is run.

**Comment by shuds13** [comment]
@RikkiButler20  I think we should just run `tutorial_calling.py` as this does not require any extra dependencies and this makes a good smoke test. `tutorial_aposmm` needs a variant. I'm wondering if this is working, as I will be doing a release of libEnsemble on Monday. Thanks for your help.

**Comment by RikkiButler20** [comment]
@shuds13 Thanks, then I'll just go with `tutorial_calling.py`
Currently this doesn't pass when running `spack test run py-libensemble`.

**Comment by shuds13** [comment]
I've also realised that the tutorial does not need the "--comms local --nworkers 4". It has that built in. But that should not make it fail. 
Can you see if the correct scripts get copied into where the tests run. I was concerned that examples is outside the python 'package'.

**Comment by shuds13** [comment]
@RikkiButler20

The only test I would run here is the `tutorial_calling`. The `tutorial_sim` and `tutorial_gen` are actually used by that, rather than standalone tests. I'm making this a bit clearer at the moment and moving tutorials into sub-directories, so I'll have to change that slightly after releasing.

If you want there to be more tests, we could use either some of the unit tests in libensemble/tests/unit_tests - these are just serial, but not complete libEnsemble runs. Or some complete libensemble parallel tests in libensemble/tests/regression_tests such as `test_uniform_sampling.py` that can be run like:

python test_uniform_sampling.py --comms local --nworkers 3

This is a good one as it does not requires any other dependencies.

I'm happy to try making these changes, but I dont want to clash. So let me know if I should go ahead?

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 1]
```diff
@@ -56,4 +57,31 @@ class PyLibensemble(PythonPackage):
     depends_on('py-mpmath', type=('build', 'run'), when='+mpmath')
     depends_on('py-deap', type=('build', 'run'), when='+deap')
     depends_on('tasmanian+python', type=('build', 'run'), when='+tasmanian')
+    depends_on('py-matplotlib', type=('run'))
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-libensemble/package.py]: Can't recall:  Is this supposed to be a `test` dependency or actually a `run` dependency?

**Comment by tldahlgren** [comment]
@RikkiButler20 Once you decide how to proceed (see shuds13's question) and are ready for it to merge, you can click the `Ready for review` button.

File: var/spack/repos/builtin/packages/py-libensemble/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/py-libensemble/package.py]: If the former, please add a TODO comment so we know it needs to be changed once the `test` dependencies are properly supported.

**Comment by shuds13** [comment]
@RikkiButler20 I propose pulling this in with the last 2 lines removed (if it is working). You could keep the branch for another future PR maybe.

I am releasing libEnsemble v0.8.0 today, so I'll try running this locally, and if it works I'll use it, otherwise, I'll release without for now.

**Comment by shuds13** [comment]
@RikkiButler20 Does this now work for you?

**Comment by tldahlgren** [comment]
> `spack install --test=root --overwrite -v -y py-libensemble`
> 
> but I get a message I don't understand
> 
> `Warning: RUN-TESTS: method not implemented [test]`
> 
> and I dont think the test is run.

The `RUN-TESTS` output is coming from the build-time Python import tests that are automatically executed.

**Comment by tldahlgren** [comment]
> This test would be run from the command line as:
> 
> `python tutorial_calling.py --comms local --nworkers 4`
> 

We generally want to avoid using multiple processes for the initial stand-alone tests since it can take a significant amount of time for resources to be allocated to the tests.  The original (current?) vision is to have multi-process tests run using another mechanism only if the initial stand-alone tests pass.

So the preference, assuming `--nworkers` is the number of processes, is to set the value to `1`.  Hopefully that would still allow the test to confirm the software still works as expected.

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 2]
```diff
@@ -57,3 +58,27 @@ class PyLibensemble(PythonPackage):
...
+
+        test_dir = join_path(self.test_suite.current_test_cache_dir,
+                             'examples', 'tutorials')
+
+        if not os.path.exists(test_dir):
+            print('Skipping {0} test'.format(exe))
+            return
+
+        self.run_test(self.spec['python'].command.path,
+                      options=[exe, '--nworker', '1'],
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-libensemble/package.py]:
```suggestion
                      options=[exe, '--nworkers', '1'],
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/py-libensemble/package.py]: Are the `'--comms'` and `'local'` options not needed?  What do they default to if not provided?

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Minor tweak(s) based on options in the comments section.

File: var/spack/repos/builtin/packages/py-libensemble/package.py [hunk 3]
```diff
@@ -57,3 +58,27 @@ class PyLibensemble(PythonPackage):
...
+
+        test_dir = join_path(self.test_suite.current_test_cache_dir,
+                             'examples', 'tutorials')
+
+        if not os.path.exists(test_dir):
+            print('Skipping {0} test'.format(exe))
+            return
+
+        self.run_test(self.spec['python'].command.path,
+                      options=[exe, '--comms', 'local', '--nworker', '1'],
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-libensemble/package.py]: The option is still plural regardless of the number of workers provided here.

```suggestion
                      options=[exe, '--comms', 'local', '--nworkers', '1'],
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]

**Comment by shuds13** [comment]
@tldahlgren 

> The `RUN-TESTS` output is coming from the build-time Python import tests that are automatically executed.

This is an error I've got trying this in the past. I can't run anything right now due to an equipment problem, but maybe I should raise this as a Spack issue? Have you successfully run tests on other python packages?

**Comment by shuds13** [comment]
As for parallelism, to do a proper libEnsemble run will always have atleast 2 processes (1 manager and 1 worker), though still very light-weight. If we want to stick strictly to serial we could run some of the unit tests. These still import all the different components (across the tests).

**Comment by shuds13** [comment]
@RikkiButler20 I will be releasing libEnsemble tomorrow. I assume this test still doesn't work, as I'm not able to run anything in Spack currently, not even "spack spec py-libensemble". So I guess this can wait for now, and we can bring in when its working.

**Comment by shuds13** [comment]
I have got spack working again. But if I run this, self.test_suite is `None`.

**Review by tldahlgren — DISMISSED** [review]
LGTM

**Comment by tldahlgren** [comment]
@RikkiButler20 Looks like there are conflicts that need to be resolved.

**Comment by shuds13** [comment]
I've released v0.8.0 so probably want to merge from develop

Also, that test has been moved into the subdirectory simple_sine.

**Comment by tldahlgren** [comment]
> I've released v0.8.0 so probably want to merge from develop
> 
> Also, that test has been moved into the subdirectory simple_sine.

FWIW I see two copies in the build stage directory I run `spack stage py-libensemble` that appear to be identical (i.e., no diffs):

```
./examples/calling_scripts/tutorial_calling.py
./examples/tutorials/simple_sine/tutorial_calling.py
```

@RikkiButler20 You could use Spack's `find` command to locate the script, e.g.:

```
exe = find(join_path(self.test_suite.current_test_cache_dir, 'examples', 'tutorials'), 'tutorial_calling.py', recursive=True)[0] 
```

to get the fully qualified path to the exe so it doesn't matter where, under the `examples/tutorials` path, the file is located.

**Comment by shuds13** [comment]
@RikkiButler20 @tldahlgren 

I've realised two things.

1. I realise I was misunderstanding the script a bit. So I understand now it is not intended to run the test with `spack install --test...`. Rather on a regular install, it copies to the test directory, and then `spack test run`... triggers the test function. This seems to work for me now, after getting round various spack problems.

2. I think I've made things difficult by putting the test in a sub-directory for the v0.8.0 release, because now that is inconsistent across versions. But I can suggest a hopefully simple solution. That is to instead use this test.

libenseble/tests/regression_tests/test_uniform_sampling.py

The last part of the file becomes:

```python
    @run_after('install')
    def cache_test_sources(self):
        """Copy the example source files after the package is installed to an
        install test subdirectory for use during `spack test run`."""
        self.cache_extra_test_sources(join_path('libensemble', 'tests', 'regression_tests'))

    def run_tutorial_tests(self, exe):
        """Run tutorials stand alone test"""

        test_dir = join_path(self.test_suite.current_test_cache_dir,
                                        'libensemble', 'tests', 'regression_tests')

        if not os.path.exists(test_dir):
            print('Skipping {0} test'.format(exe))
            return

        self.run_test(self.spec['python'].command.path,
                      options=[exe, '--comms', 'local', '--nworkers', '1'],
                      purpose='test: run {0} example'.format(exe),
                      work_dir=test_dir)

    def test(self):
        self.run_tutorial_tests('test_uniform_sampling.py')
```

This seemed to work for me! 

Also, if more than one test is desirable, we could run others from that directory, like test_fast_alloc.py could be run with the same configuration.

**Comment by shuds13** [comment]
@RikkiButler20 @tldahlgren This now works for me without the matplotlib line, so maybe we can remove that? Then I'm hapy to pull this in.

Edit: Actually I should say it works for recent releases, but I'm not sure the links in calling_scripts go back that far.

Edit: Correcting this again, if I go to v0.6.0, the dir exists, but the test had a different name! As I dont tihnk there is any good reason for people to install old versions, perhaps the skip line could be.

if not os.path.exists(join_path(test_dir, exe)):

That then stops me getting an error for old installs, but I dont see the 'skipping' message. I presume its getting redirected to a log somewhere.

**Comment by shuds13** [comment]
@RikkiButler20 @tldahlgren
I approve this PR

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**